ejenti/server_jobs/dashboard_data_generator.py

```diff
@@ -0,0 +1,402 @@
+                'outerRadius': '100%',
+                'shape': 'arc'
+            }
+        },
+        'tooltip': {
+            'enabled': False
+        },
+        'yAxis': {
+            'stops': [
+                [0.1, '#ff0000'],
+                [0.5, '#ffff00'],
+                [0.9, '#00ff00']
+            ],
+            'lineWidth': 0,
+            'minorTickInterval': None,
+            'tickAmount': 0,
+            'tickPixelInterval': 0,
+            'labels': {
+                'enabled': False
+            },
+            'min': 0,
+            'max': 100,
+            'tickInterval': 0.01,
+            'tickWidth': 0
+        },
+        'plotOptions': {
+            'solidgauge': {
+                'dataLabels': {
+                    'y': 5,
+                    'borderWidth': 0,
+                    'useHTML': True
+                }
+            }
+        },
+        'credits': {
+            'enabled': False
+        },
+        'series': []
+    }
+
+    JS_GAUGE_SERIES_OBJ_TEMPLATE = {
+        'name': '',
+        'data': [],
+        'dataLabels': {
+            'format': '<div style="text-align:center"><span style="font-size:25px;color:silver">{y}%</span></div>',
+        },
+        'tooltip': {
+            'valueSuffix': ''
+        }
+    }
+
+    def __init__(self):
+        self.source_data = []
+        self.casename_set = set()
+        self.latest_timestamp = ''
+
+        self._generate_source_data()
+
+    def _generate_source_data(self):
+        table_obj = GenerateBackfillTableHelper.get_history_archive_perfherder_relational_table(
+            DashboardDataGenerator.PLATFORM)
+
+        ret_list = []
+
+        if table_obj:
+            self.latest_timestamp = sorted(table_obj.keys())[-1]
+            self.latest_timestamp_js = int(self.latest_timestamp) * 1000
+            # get 28 builds (almost 14 days) timestamp
+            latest_28_builds_timestamp_list = sorted(table_obj.keys())[-28:]
+
+            for timestamp in latest_28_builds_timestamp_list:
+                revison = table_obj.get(timestamp, {}).get(DashboardDataGenerator.KEY_REVISON, '')
+                perf_data_obj_dict = table_obj.get(timestamp, {}).get(DashboardDataGenerator.KEY_PERFHERDER_DATA, {})
+
+                for key, value in perf_data_obj_dict.items():
+                    ret_obj = {}
+
+                    # ex: amazon_ail_type_in_search_field:firefox:windows8-64
+                    casename, browser, platform = key.split(':')
+
+                    self.casename_set.add(casename)
+
+                    value_list = value.get(DashboardDataGenerator.KEY_VLAUE)
+                    signature = value.get(DashboardDataGenerator.KEY_SIGNATURE)
+
+                    ret_obj[DashboardDataGenerator.KEY_TIMESTAMP] = timestamp
+                    ret_obj[DashboardDataGenerator.KEY_TIMESTAMP_JS] = int(timestamp) * 1000
+                    ret_obj[DashboardDataGenerator.KEY_CASENAME] = casename
```

Comment:
When there is missing result at latest build, the overall progress will be incorrect.
Should adding a set for storing all `case name` for reference at `generate_latest_build_overall_progress_for_platform()`.